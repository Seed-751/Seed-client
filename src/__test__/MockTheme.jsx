import React from "react";
import PropTypes from "prop-types";
import { ThemeProvider } from "styled-components";
import GrobalStyles from "../styles/index";
import theme from "../styles/theme";

export default function MockTheme({ children }) {
  return (
    <ThemeProvider theme={theme}>
      <GrobalStyles />
      {children}
    </ThemeProvider>
  );
}

MockTheme.propTypes = {
  children: PropTypes.element.isRequired,
};
