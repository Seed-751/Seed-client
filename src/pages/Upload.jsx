import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import { useDropzone } from "react-dropzone";
import styled from "styled-components";

import { useForm } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";

import Input from "../components/shared/Input";
import Button from "../components/shared/Button";
import requestUploadMusic from "../api/requestUploadMusic";
import validateMetaData from "../utils/validateMetaData";
import { INITIAL_PREVIEW_IMAGE, ERROR, GENRE_OPTIONS } from "../constants";

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 5%;
  text-align: center;

  div > * {
    display: block;
    width: 400px;
  }

  strong {
    font-size: 24px;
    font-weight: 600;
    letter-spacing: 0.06em;
  }

  p {
    margin-top: 28px;
    font-size: 14px;
    color: ${({ theme }) => theme.color.gray};
  }
`;

const InputBox = styled.div`
  margin-top: 10px;
  margin-Bottom: 10px;
  width: 100%;
  text-align: left;

  label {
    margin-left: 15px;
  }

  p {
    margin: 0;
    margin-left: 15px;
    color: red;
  }
`;

const UploadInput = styled(Input)`
  border-radius: 5px;
`;

const UploadImageInput = styled.input`
  outline: none;
  border-style: dashed;
  border-color: ${({ theme }) => theme.color.gray};
  border-width: 2px;
  border-radius: 2px;
`;

const UploadTextarea = styled.textarea`
  resize: vertical;
`;

const PreviewBox = styled.div`
  height: 100%;
  border: 1px solid black;

  img {
    height: 100%;
  }
`;

const UploadAudioInput = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  border-width: 2px;
  border-radius: 2px;
  border-color: ${({ theme }) => theme.color.gray};
  border-style: dashed;
  background-color: ${({ theme }) => theme.color.white};
  outline: none;
  transition: border .24s ease-in-out;

  &:hover {
    cursor: pointer;
  }
`;

export default function Upload() {
  const [previewImage, setPreviewImage] = useState(INITIAL_PREVIEW_IMAGE);
  const [audioError, setAudioError] = useState(null);
  const history = useHistory();
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();
  const {
    getRootProps,
    getInputProps,
    acceptedFiles,
  } = useDropzone({
    accept: "audio/*",
    onDrop,
    multiple: true,
    validator: validateAudio,
  });

  async function onDrop(acceptedFiles) {
    if (!acceptedFiles.length) {
      return setAudioError([ERROR.inputAudioFile]);
    }

    await Promise.all(acceptedFiles.map((file) => {
      return validateMetaData(file);
    })).then((result) => setAudioError(result));
  }

  async function handleUploadMusic(data) {
    const isNotError = audioError.every((error) => {
      return error === null;
    });

    if (!isNotError) {
      return setAudioError([ERROR.checkAudioFile]);
    }

    const { image } = data;
    const musicInfo = {
      ...data,
      image: image[0],
      audioFiles: acceptedFiles,
    };

    try {
      const result = await requestUploadMusic(musicInfo);

      if (result.success) {
        return history.push("/");
      }

    } catch (err) {
      console.log(err);
    }
  }

  function handleImage(e) {
    e.target.files[0]
      ? setPreviewImage(URL.createObjectURL(e.target.files[0]))
      : setPreviewImage(INITIAL_PREVIEW_IMAGE);
  }

  function validateAudio(file) {
    const type = file.type.split("/")[0];

    if (type !== "audio") {
      return ERROR.inputAudioFile;
    }
  }

  return (
    <Container>
      <div>
        <h1>Upload Music</h1>
        <form
          onSubmit={handleSubmit(handleUploadMusic)}
          encType="multipart/form-data"
        >

          <InputBox>
            <label>Title</label>
            <UploadInput
              name="title"
              {...register("title", {
                required: ERROR.inputTitle,
              })}
            />
            <ErrorMessage
              errors={errors}
              name="title"
              render={({ message }) => <p>{message}</p>}
            />
          </InputBox>

          <InputBox>
            <label>Image</label>
            <PreviewBox>
              <img src={previewImage} />
            </PreviewBox>
            <UploadImageInput
              name="image"
              type="file"
              {...register("image", {
                onChange: handleImage,
                required: ERROR.inputImage,
              })}
            />
            <ErrorMessage
              errors={errors}
              name="image"
              render={({ message }) => <p>{message}</p>}
            />
          </InputBox>

          <InputBox>
            <label>Audio</label>
            <UploadAudioInput {...getRootProps()}>
              <input
                {...getInputProps()}
              />
              <p style={{ color: "gray" }}>Drag and drop audio files here, or click to select files</p>
            </UploadAudioInput>
            <ul>
              {acceptedFiles?.map((file) => (
                <li key={file.path}>{file.path}</li>
              ))}
            </ul>
            {audioError && <p>{audioError}</p>}
          </InputBox>

          <InputBox>
            <label>Genre</label>
            <select
              name="genre"
              {...register("genre", {
                required: ERROR.inputGenre,
              })}
            >
              {GENRE_OPTIONS.map((genre) => (
                <option key={genre} value={genre}>{genre}</option>
              ))}
            </select>
            <ErrorMessage
              errors={errors}
              name="genre"
              render={({ message }) => <p>{message}</p>}
            />
          </InputBox>

          <InputBox>
            <label>Description</label>
            <UploadTextarea
              name="description"
              {...register("description", {
                required: ERROR.inputDescription,
              })}
            />
            <ErrorMessage
              errors={errors}
              name="description"
              render={({ message }) => <p>{message}</p>}
            />
          </InputBox>

          <Button type="submit">Upload</Button>
        </form>
      </div>
    </Container>
  );
}
