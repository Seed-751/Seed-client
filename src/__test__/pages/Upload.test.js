import React from "react";
import { render } from "@testing-library/react";

import Upload from "../../pages/Upload";
import MockTheme from "../MockTheme";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
}));

describe("Upload page", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <MockTheme>
        <Upload />
      </MockTheme>
    );

    expect(container).toMatchSnapshot();
  });
});
