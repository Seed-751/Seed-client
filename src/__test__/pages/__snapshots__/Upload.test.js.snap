// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload page matches snapshot 1`] = `
<div>
  <div
    class="sc-iqHYmW eLLWmM"
  >
    <div
      class="sc-crrszt fEIOER"
    >
      <strong>
        Upload Music
      </strong>
      <form
        class="sc-dlfnuX kA-Dyrb"
        enctype="multipart/form-data"
      >
        <div
          class="sc-hKgJUU iTBZof"
        >
          <div
            class="sc-jrAFXE hvXkzg"
          >
            <div
              class="sc-eCstlR eQcvTH"
            >
              <label
                id="image-label"
              >
                Image
              </label>
              <div
                class="sc-fubCzh kAJbsJ"
              >
                <img
                  src="/defaultImage.png"
                />
              </div>
              <input
                aria-labelledby="image-label"
                class="sc-gKseQn djnPYe"
                name="image"
                type="file"
              />
            </div>
            <div
              class="sc-eCstlR eQcvTH"
            >
              <label
                id="audio-label"
              >
                Audio
              </label>
              <div
                class="sc-pGacB EciOV"
                tabindex="0"
              >
                <input
                  accept="audio/*"
                  aria-labelledby="audio-label"
                  autocomplete="off"
                  multiple=""
                  style="display: none;"
                  tabindex="-1"
                  type="file"
                />
                <p
                  style="color: gray;"
                >
                  Drag and drop audio files here, or click to select files
                </p>
              </div>
              <div
                class="files-container"
              >
                <p>
                  Accepted Files
                </p>
              </div>
              <div
                class="files-container"
              >
                <p>
                  Rejected Files
                </p>
              </div>
            </div>
          </div>
          <div
            class="sc-kEjbQP jvxEBd"
          >
            <div
              class="sc-eCstlR eQcvTH"
            >
              <label
                id="title-label"
              >
                Title
              </label>
              <input
                aria-labelledby="title-label"
                class="sc-bdfBQB sc-jSgvzq gpSnii kNqZOj"
                name="title"
              />
            </div>
            <div
              class="sc-eCstlR eQcvTH"
            >
              <label
                id="genre-label"
              >
                Genre
              </label>
              <select
                aria-labelledby="genre-label"
                name="genre"
              >
                <option
                  value="classic"
                >
                  classic
                </option>
                <option
                  value="jazz"
                >
                  jazz
                </option>
                <option
                  value="pop"
                >
                  pop
                </option>
                <option
                  value="rock"
                >
                  rock
                </option>
                <option
                  value="hip hop"
                >
                  hip hop
                </option>
                <option
                  value="rhythm and blues"
                >
                  rhythm and blues
                </option>
                <option
                  value="reggae"
                >
                  reggae
                </option>
                <option
                  value="funk"
                >
                  funk
                </option>
                <option
                  value="disco"
                >
                  disco
                </option>
                <option
                  value="electronic music"
                >
                  electronic music
                </option>
                <option
                  value="vocal music"
                >
                  vocal music
                </option>
                <option
                  value="ballad"
                >
                  ballad
                </option>
              </select>
            </div>
            <div
              class="sc-eCstlR eQcvTH"
            >
              <label
                id="description-label"
              >
                Description
              </label>
              <textarea
                aria-labelledby="description-label"
                class="sc-iBPTik fRXNcT"
                name="description"
              />
            </div>
          </div>
        </div>
        <button
          class="sc-gsTEea bzwDJA submit"
          type="submit"
        >
          Upload
        </button>
      </form>
    </div>
  </div>
</div>
`;
